Decrease use of StringView::character8() / character16()

cdumez opened the pull request:
#### 4a5f7f431122a5e13b4d81f3cf9d4352e1338d98
<pre>
Decrease use of StringView::character8() / character16()
<a href="https://bugs.webkit.org/show_bug.cgi?id=273040">https://bugs.webkit.org/show_bug.cgi?id=273040</a>

Reviewed by Darin Adler.

Decrease use of StringView::character8() / character16() in preparation
for dropping them.

* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::Parser):
* Source/WTF/wtf/JSONValues.cpp:
(WTF::JSONImpl::Value::parseJSON):
* Source/WTF/wtf/text/AdaptiveStringSearcher.h:
(WTF::searchStringRaw): Deleted.
* Source/WTF/wtf/text/Base64.cpp:
(WTF::base64Decode):
(WTF::base64DecodeToString):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::find):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::find const):
(WTF::getCharactersWithASCIICaseInternal):
(WTF::StringView::getCharactersWithASCIICase const):
(WTF::normalizedNFC):
(WTF::makeStringByReplacingAll):
(WTF::makeStringBySimplifyingNewLinesSlowCase):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::span8 const):
(WTF::StringView::span16 const):
(WTF::StringView::span&lt;LChar&gt; const):
(WTF::StringView::span&lt;UChar&gt; const):
(WTF::StringView::characterAt const):
(WTF::StringView::unsafeCharacterAt const):
(WTF::equal):
(WTF::StringView::CodePoints::Iterator::Iterator):
(WTF::StringView::trim const):
(WTF::startsWith):
(WTF::startsWithIgnoringASCIICase):
(WTF::endsWith):
(WTF::endsWithIgnoringASCIICase):
(WTF::StringView::characters8 const): Deleted.
(WTF::StringView::characters16 const): Deleted.
(WTF::StringView::characters&lt;LChar&gt; const): Deleted.
(WTF::StringView::characters&lt;UChar&gt; const): Deleted.
* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::setTextForIterator):
(WTF::numGraphemeClusters):
(WTF::numCodeUnitsInGraphemeClusters):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::span&lt;LChar&gt; const):
(WTF::String::span&lt;UChar&gt; const):
* Source/WTF/wtf/text/cf/StringViewCF.cpp:
(WTF::StringView::createCFString const):
(WTF::StringView::createCFStringWithoutCopying const):
* Source/WTF/wtf/text/cocoa/StringViewCocoa.mm:
(WTF::StringView::createNSString const):
(WTF::StringView::createNSStringWithoutCopying const):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::setText):
* Source/WTF/wtf/text/icu/UTextProvider.h:
(WTF::initializeContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::openLatin1ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.h:
* Source/WTF/wtf/text/icu/UTextProviderUTF16.cpp:
(WTF::openUTF16ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderUTF16.h:
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
(WTF::createIterator):
* Source/WebCore/dom/SpaceSplitString.cpp:
(WebCore::spaceSplitStringContainsValueInternal):

Canonical link: <a href="https://commits.webkit.org/277796@main">https://commits.webkit.org/277796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b0645298eb5d18b6c2f9caf89a146da28b663c2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48614 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27825 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51301 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44679 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50919 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25355 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39756 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25491 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41944 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20853 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43122 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6670 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41913 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44907 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43607 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53209 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/48105 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23661 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19959 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47048 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24927 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42149 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25731 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55599 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6925 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24648 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11434 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/Base64.cpp:
```diff
@@ -263,10 +263,9 @@ std::optional<Vector<uint8_t>> base64Decode(std::span<const std::byte> input, Ba
-    if (!length || input.is8Bit())
-        return base64DecodeInternal(std::span(input.characters8(), length), mode);
-    return base64DecodeInternal(std::span(input.characters16(), length), mode);
+    if (input.isEmpty() || input.is8Bit())
```
I think this only needs to check `is8Bit`. I don’t think `isEmpty` leads to additional optimization or is needed for correctness.

darinadler commented on Source/WTF/wtf/text/Base64.cpp:
```diff
@@ -277,10 +276,9 @@ String base64DecodeToString(StringView input, Base64DecodeMode mode)\
-    if (!length || input.is8Bit())
-        return toString(base64DecodeInternal<LChar, StringImplMalloc>(std::span(input.characters8(), length), mode));
-    return toString(base64DecodeInternal<UChar, StringImplMalloc>(std::span(input.characters16(), length), mode));
+    if (input.isEmpty() || input.is8Bit())
```
I think this only needs to check `is8Bit`. I don’t think `isEmpty` leads to additional optimization or is needed for correctness.

darinadler commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -902,12 +902,12 @@ size_t StringImpl::find(StringView matchString)
         if (is8Bit()) {
             if (matchString.is8Bit())
-                return WTF::find(span8(), matchString.characters8()[0]);
-            return WTF::find(span8(), matchString.characters16()[0]);
+                return WTF::find(span8(), matchString.span8().front());
+            return WTF::find(span8(), matchString.span16().front());
```
This seems unnecessary. Why not just `matchString[0]`? Is this somehow more optimal?

weinig replied:
I think this is likely a bit more optimal, but the compiler is likely optimizing out the issue in practice.

matchString[0] (which will call matchString.characterAt(0)) will re-check is8Bit() and won't return the right type for the 8bit case.

darinadler commented on Source/WTF/wtf/text/StringImpl.cpp:
```diff
@@ -902,12 +902,12 @@ size_t StringImpl::find(StringView matchString)
         }
         if (matchString.is8Bit())
-            return WTF::find(span16(), matchString.characters8()[0]);
-        return WTF::find(span16(), matchString.characters16()[0]);
+            return WTF::find(span16(), matchString.span8().front());
+        return WTF::find(span16(), matchString.span16().front());
```
This seems unnecessary. Why not just `matchString[0]`? Is this somehow more optimal?

darinadler commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -586,18 +586,17 @@ inline StringView StringView::substring(unsigned start, unsigned length) const
     if (is8Bit())
-        return characters8()[index];
-    return characters16()[index];
+        return span8().data()[index];
+    return span16().data()[index];
```
We should consider getting rid of `unsafeCharacterAt` unless it’s a measurable performance win, no?

cdumez replied:
It is private and only used from iterator code so it's not terrible. But yes, we should try and get rid of it if there is no performance impact. I'll do separately though as this requires further perf testing.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -107,6 +107,7 @@ class String final {
 
     // Return characters8() or characters16() depending on CharacterType.
     template<typename CharacterType> const CharacterType* characters() const;
+    template<typename CharacterType> std::span<const CharacterType> span() const;
```
Eventually we should make our classes consistent and remove synonyms. I wonder if we should standardize on `span<const LChar>`, `span<LChar>`, or `span8`?

darinadler commented on Source/WTF/wtf/text/icu/UTextProvider.h:
```diff
@@ -44,13 +44,14 @@ inline UTextProviderContext uTextProviderContext(const UText* text, int64_t nati
-    text->p = string;
-    text->a = length;
+    text->context = static_cast<const void*>(string.data());
+    text->p = static_cast<const void*>(string.data());
```
If we are actually assigning to a `const void*`, then we don’t need the cast.

cdumez committed:
Decrease use of StringView::character8() / character16()
https://bugs.webkit.org/show_bug.cgi?id=273040

Reviewed by Darin Adler.

Decrease use of StringView::character8() / character16() in preparation
for dropping them.

* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::Parser):
* Source/WTF/wtf/JSONValues.cpp:
(WTF::JSONImpl::Value::parseJSON):
* Source/WTF/wtf/text/AdaptiveStringSearcher.h:
(WTF::searchStringRaw): Deleted.
* Source/WTF/wtf/text/Base64.cpp:
(WTF::base64Decode):
(WTF::base64DecodeToString):
* Source/WTF/wtf/text/StringCommon.h:
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::find):
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::find const):
(WTF::getCharactersWithASCIICaseInternal):
(WTF::StringView::getCharactersWithASCIICase const):
(WTF::normalizedNFC):
(WTF::makeStringByReplacingAll):
(WTF::makeStringBySimplifyingNewLinesSlowCase):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::span8 const):
(WTF::StringView::span16 const):
(WTF::StringView::span<LChar> const):
(WTF::StringView::span<UChar> const):
(WTF::StringView::characterAt const):
(WTF::StringView::unsafeCharacterAt const):
(WTF::equal):
(WTF::StringView::CodePoints::Iterator::Iterator):
(WTF::StringView::trim const):
(WTF::startsWith):
(WTF::startsWithIgnoringASCIICase):
(WTF::endsWith):
(WTF::endsWithIgnoringASCIICase):
(WTF::StringView::characters8 const): Deleted.
(WTF::StringView::characters16 const): Deleted.
(WTF::StringView::characters<LChar> const): Deleted.
(WTF::StringView::characters<UChar> const): Deleted.
* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::setTextForIterator):
(WTF::numGraphemeClusters):
(WTF::numCodeUnitsInGraphemeClusters):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::span<LChar> const):
(WTF::String::span<UChar> const):
* Source/WTF/wtf/text/cf/StringViewCF.cpp:
(WTF::StringView::createCFString const):
(WTF::StringView::createCFStringWithoutCopying const):
* Source/WTF/wtf/text/cocoa/StringViewCocoa.mm:
(WTF::StringView::createNSString const):
(WTF::StringView::createNSStringWithoutCopying const):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::setText):
* Source/WTF/wtf/text/icu/UTextProvider.h:
(WTF::initializeContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::openLatin1ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderLatin1.h:
* Source/WTF/wtf/text/icu/UTextProviderUTF16.cpp:
(WTF::openUTF16ContextAwareUTextProvider):
* Source/WTF/wtf/text/icu/UTextProviderUTF16.h:
* Source/WTF/wtf/unicode/icu/CollatorICU.cpp:
(WTF::createIterator):
* Source/WebCore/dom/SpaceSplitString.cpp:
(WebCore::spaceSplitStringContainsValueInternal):

Canonical link: https://commits.webkit.org/277796@main
